common/src/main/java/io/netty/util/CachingDomainNameMapping.java

```diff
@@ -0,0 +1,256 @@
+         * to be thrown
+         *
+         * @return new {@link DomainNameMapping} instance
+         */
+        @SuppressWarnings("unchecked")
+        public CachingDomainNameMapping<V> build() {
+            return new CachingDomainNameMapping<V>(defaultValue, new LinkedHashMap<String, V>(exactMatchMap),
+                    wildcardMap.entrySet().toArray(new Map.Entry[0]), cacheCapacity);
+        }
+    }
+
+    /**
+     * Cache that uses "relaxed" concurrency access, which means different threads may see updates done by other
+     * threads but are not guaranteed to do so. This is OK as for a given key we will always cache the same value so
+     * at worse we will do some extra work.
+     *
+     * The implementation idea is based on
+     * <a href="http://vanillajava.blogspot.com/2014/12/on-heap-vs-off-heap-memory-usage.html">On heap vs off
+     * heap memory usage</a>.
+     *
+     * @param <V> the type of key to cache.
+     */
+    private static final class RelaxedCache<V> {
+
+        private static final class CacheEntry<V> {
+            final String key;
+            final V value;
+
+            CacheEntry(String key, V value) {
+                this.key = key;
+                this.value = value;
+            }
+        }
+
+        private final CacheEntry<V>[] entries;
+        private final int shift;
+
+        @SuppressWarnings("unchecked")
+        RelaxedCache(final int capacity) {
+            entries = (CacheEntry<V>[]) new Object[MathUtil.findNextPositivePowerOfTwo(capacity)];
+            //log2 of entries.length
+            shift = 31 - Integer.numberOfLeadingZeros(entries.length);
+        }
+
+        // fast % operation with power of 2 entries.length
+        private int firstIndex(int hashCode) {
+            return hashCode & (entries.length - 1);
+        }
+
+        private int secondIndex(int hashCode) {
+            return (hashCode >> shift) & (entries.length - 1);
+        }
+
+        private V value(int idx, String key) {
+            final CacheEntry<V> entry = entries[idx];
+            if (entry != null && entry.key.equals(key)) {
+                return entry.value;
+            }
+            return null;
+        }
+
+        V get(final String key) {
+            final int hashCode = key.hashCode();
+            final int firstIndex = firstIndex(hashCode);
+            final V firstValue = value(firstIndex, key);
+            return firstValue != null ? firstValue :
+                    value(secondIndex(hashCode), key);
+        }
+
+        void put(String key, V value) {
+            CacheEntry<V> entry = new CacheEntry<V>(key, value);
+            final int hashCode = key.hashCode();
+            final int firstIndex = firstIndex(hashCode);
+            final V firstValue = value(firstIndex, key);
+            if (firstValue == null) {
+                entries[firstIndex] = entry;
+            } else {
+                entries[secondIndex(hashCode)] = entry;
```

Comment:
@franz1981 I wonder if we should maybe use something like `ThreadLocalRandom.nextBoolean()` to decide if we put it in the `firstIndex`  vs `secondIndex` if both `indices` are already full. WDYT?

Reply:
@franz1981 basically:

```
diff --git a/common/src/main/java/io/netty/util/CachingDomainNameMapping.java b/common/src/main/java/io/netty/util/CachingDomainNameMapping.java
index 6f87532e4e..572205d27f 100644
--- a/common/src/main/java/io/netty/util/CachingDomainNameMapping.java
+++ b/common/src/main/java/io/netty/util/CachingDomainNameMapping.java
@@ -17,6 +17,7 @@ package io.netty.util;

 import io.netty.util.internal.MathUtil;
 import io.netty.util.internal.StringUtil;
+import io.netty.util.internal.ThreadLocalRandom;

 import java.util.Collections;
 import java.util.LinkedHashMap;
@@ -249,7 +250,12 @@ public final class CachingDomainNameMapping<V> extends DomainNameMapping<V> {
             if (firstValue == null) {
                 entries[firstIndex] = entry;
             } else {
-                entries[secondIndex(hashCode)] = entry;
+                int secondIndex = secondIndex(hashCode);
+                if (entries[secondIndex] != null && ThreadLocalRandom.current().nextBoolean()) {
+                    entries[firstIndex] = entry;
+                } else {
+                    entries[secondIndex] = entry;
+                }
             }
         }
     }
```

Reply:
Can be done TBH and so can choose to move linearly until a certain distance (effectively a linear probing) but with a limited horizon too

Reply:
@franz1981 so yay or nay ?

Reply:
yay! :+1: